### PR TITLE
Improve reliability of async indexing while the user is editing the file, implement the memory usage optimization for the indexer, make its speed configurable

### PR DESCRIPTION
**Pull request by dmitmel**
Ref: #12 \
Depends on: #22

This PR implements the main memory usage optimization. Also I hope that the documentation I have added is enough to justify advanced options like this.

### CODE REVIEW

**Comment by dmitmel** [comment]
This is the last remaining part of #12 left. I'm working on a few bugs related to editing while the buffer is being indexed.

**Comment by dmitmel** [comment]
@hrsh7th This PR is ready for review. Sorry for taking so long.

File: lua/cmp_buffer/timer.lua [hunk 1]
```diff
@@ -0,0 +1,68 @@
+---This timer matches the semantics of setInterval and clearInterval of
```
Comment (hrsh7th) [inline lua/cmp_buffer/timer.lua:1]: Good abstraction.

File: lua/cmp_buffer/buffer.lua [hunk 2]
```diff
@@ -112,48 +107,67 @@ function buffer.safe_buf_call(self, callback)
...
-        end
-      end)
-      chunk_start = chunk_end
-      self:mark_all_lines_dirty()
-      self.words_distances_dirty = true
+    local batch_start = self.timer_current_line
+    local batch_size = self.opts.indexing_batch_size
+    -- NOTE: self.lines_count may be modified by the indexer.
+    local batch_end = batch_size >= 1 and math.min(batch_start + batch_size, self.lines_count) or self.lines_count
+    if batch_end >= self.lines_count then
```
Comment (hrsh7th) [inline lua/cmp_buffer/buffer.lua:157]: I think this line should move to the end of timer callback.

File: lua/cmp_buffer/buffer.lua [hunk 3]
```diff
@@ -238,23 +271,13 @@ function buffer.watch(self)
...
-          self.lines_words[i] = {}
-        end
-      elseif new_lines_count < self.lines_count then -- remove
-        for i = self.lines_count, new_lines_count + 1, -1 do
-          self.lines_words[i] = nil
-        end
+      -- clear all lines
+      for i = self.lines_count, 1, -1 do
+        self.lines_words[i] = nil
       end
```
Comment (hrsh7th) [inline lua/cmp_buffer/buffer.lua:277]: use clear_table instead.

File: lua/cmp_buffer/buffer.lua [hunk 4]
```diff
@@ -218,6 +232,25 @@ function buffer.watch(self)
...
+          -- synchronously index this area it in a moment, so the indexer
+          -- should resume from right after the edit range.
+          self.timer_current_line = new_last_line
+        elseif self.timer_current_line >= old_last_line then
+          -- The indexer was somewhere past the current text edit. This means
+          -- that the line numbers could have changed, and the indexing
+          -- position must be adjusted accordingly.
+          self.timer_current_line = self.timer_current_line + delta
+        end
+      end
```
Comment (hrsh7th) [inline lua/cmp_buffer/buffer.lua:252]: smart!

File: lua/cmp_buffer/buffer.lua (hunk 2 above)
Comment (hrsh7th) [reply lua/cmp_buffer/buffer.lua:157]: I'll merge this if this code is in the correct position.

Comment (dmitmel) [reply lua/cmp_buffer/buffer.lua:163]: This is intentional, it is a defensive style. Basically, if an error occurs within `index_range`, the indexer will not lock up in an infinite loop (like it was in #29) where it just tries to re-index the same group of lines again and again, always throwing an error which prevents it from advancing further in the file. I have changed this in the latest commit in such a way that `self:index_range()` is the very last statement in both this callback and the `on_lines` watcher.

**Comment by hrsh7th** [comment]
LGTM!

**Pull request merged**